modules/gapi/src/compiler/passes/pattern_matching.cpp

```diff
@@ -0,0 +1,371 @@
+                    matchedVisitedFirstDataNodes.clear();
+                    inputApiMatch.clear();
+                    matched = false;
+                    break;
+                }
+                inputApiMatch[(*patternIt)->srcNode()] = (*matchedIt)->srcNode();
+            }
+
+        }
+
+        if (matched) {
+            std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> visitedLastDataNodes;
+            for (auto it = subgraphOuts.begin(); it != subgraphOuts.end() && matched; ++it) {
+                auto match = *it;
+                auto patternOututEdges = match.first->outEdges();
+                auto compOutputEdges = match.second->outEdges();
+
+                if (match.first->outNodes().size() != match.second->outNodes().size()) {
+                    visitedLastDataNodes.clear();
+                    outputApiMatch.clear();
+                    matched = false;
+                    break;
+                }
+
+                for (auto patternIt = patternOututEdges.begin(); patternIt != patternOututEdges.end(); ++patternIt) {
+
+                    if ((*patternIt)->dstNode()->outEdges().size() != 0) {
+                        continue;
+                    }
+
+                    auto matchedIt = std::find_if(compOutputEdges.begin(), compOutputEdges.end(),
+                        [&patternIt, &patternGraph, &compGraph, &visitedLastDataNodes](const ade::EdgeHandle& compEdge) -> bool {
+                        auto patternOutputPort = patternGraph.metadata(*patternIt).get<cv::gimpl::Output>().port;
+                        auto compOutputPort = compGraph.metadata(compEdge).get<cv::gimpl::Output>().port;
+
+                        if (patternOutputPort != compOutputPort) {
+                            return false;
+                        }
+
+                        // Not sure that it is needed at all, we can't have such case with multiple outputs to 1 data node
+                        auto foundit = std::find_if(visitedLastDataNodes.begin(), visitedLastDataNodes.end(), [&compEdge](const ade::NodeHandle& visitedNode) { return compEdge->dstNode() == visitedNode; });
+                        if (foundit != visitedLastDataNodes.end()) {
+                            return false;
+                        }
+
+                        visitedLastDataNodes.insert(compEdge->dstNode());
+
+                        return true;
+                    });
+
+                    if (matchedIt == compOutputEdges.end()) {
+                        visitedLastDataNodes.clear();
+                        outputApiMatch.clear();
+                        matched = false;
+                        break;
+                    }
+                    outputApiMatch[(*patternIt)->dstNode()] = (*matchedIt)->dstNode();
+                }
+
+            }
+        }
+    }
+
+    SubgraphMatch subgraph{};
+
+    if (notFound || !matched) {
+        return subgraph;
+    }
+
+    subgraph.inputDataNodesMatches = inputApiMatch;
```

Comment:
I suggest to use move semantics here: I guess `std::unordered_map` filled with data can get quite "heavy"

Reply:
Done